### ISSUE TITLE
How was this made?

### ISSUE THREAD

**Issue by CJ5-Z** [PathOfBuildingCommunity/PathOfBuilding#2840]
Hello, and apologize in advance for this very amateur question!
I am just getting into building windows applications, and PoB has been a big inspiration for this. The only one I have built so far was in WinForms, and that seems very inefficient and after a quick look-through of the repository I could tell that PoB was definitely not made in this way. My question just is, how, in general, was this made? I saw some .sln files in the PoB repos so that leads me to believe it was made primarily in Visual Studio, but in what way? Again sorry for the very simple and general question but I just thought I would go here to maybe get some answers and find a starting point for myself.
Thank you very much!

**Comment by ppoelzl** [comment]
That's a very good question! I hope I can provide a good overview.

Path of Building's application logic is mostly in the Lua files you can find in this repository.
The main executable itself (we call it launcher) is in here: https://github.com/PathOfBuildingCommunity/PathOfBuilding-Launcher.
This is basically just to provide a Lua host (a Lua runtime, LuaJIT in out case), set up some environment for Lua to use, load SimpleGraphic and finally load a Lua script as main application entry point.

SimpleGraphic https://github.com/PathOfBuildingCommunity/PathOfBuilding-SimpleGraphic is our custom low level graphics library. It's based on raw OpenGL primitives, drawn in immediate mode. It exposes many functions you can call from Lua code.
There are many classes for common UI elements available, but you can also always draw with SimpleGraphic directly as well.

While Path of Building is a Lua project, the launcher and SimpleGraphic are C++ projects. You're right that you'll need Visual studio to work with the latter two.

Now if you ask me _why_ it was made this way, I can't give you a complete answer. This project is a fork of https://github.com/Openarl/PathOfBuilding, and these architecture choices were already made when we started maintaining Path of Building.

In short, Path of Building was made mostly from scratch, using Lua scripts , a C++ Lua host and a custom C++ graphics library.

If there's anything else you're interesed in, be sure to ask!

**Comment by CJ5-Z** [comment]
Thank you very much, I will definitely check all of that out! Something else I thought was interesting was the way the updater works, where it is able to get a list of the changes and output them to a window. How did you get that to work?

**Comment by ppoelzl** [comment]
It's easy, all due to the manifest file: https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/master/manifest.xml
When PoB checks for updates, it downloads this remote manifest and rehashes all application files to find the ones that need to be updated. The main logic is in https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/master/src/UpdateCheck.lua

**Comment by CJ5-Z** [comment]
Awesome, I think I'll get started experimenting around. Thanks again for your help.
